packages/codemirror-blocks/src/state/actions.ts

```diff
@@ -79,10 +79,64 @@ export const disableQuarantine = () => ({
   type: "DISABLE_QUARANTINE" as const,
 });
 
-export const setBlockMode = (enabled: boolean) => ({
-  type: "SET_BLOCK_MODE" as const,
-  enabled,
-});
+export const setBlockMode =
+  (
+    enabled: boolean,
+    editor: CMBEditor,
+    language: Language
+  ): AppThunk<ReturnType<typeof checkASTRoundtripConversion>> =>
+  (dispatch) => {
+    const result = checkASTRoundtripConversion(editor.getValue(), language);
+    if (result.successful) {
+      dispatch({
+        type: "SET_BLOCK_MODE" as const,
+        enabled,
+      });
+    }
+    return result;
+  };
+
+/**
+ * Check whether the given code can be parsed into an AST and then
+ * pretty printed back into code again, and then back into an AST again.
+ *
+ * @param oldCode text content of the code to parse
+ * @param language language to use to parse the code
+ * @returns returns a result object with the following properties:
+ *   - `oldAst`: the ast generated by parsing `oldCode`
+ *   - `newCode`: the code generated by pretty printing the old ast
+ *   - `newAst`: the ast generated by parsing the pretty printed code
+ */
+function checkASTRoundtripConversion(
+  oldCode: string,
+  language: Language
+): Result<{ oldAst: AST; newAst: AST; newCode: string }, string> {
+  let oldAst: AST;
+  try {
+    oldAst = AST.from(language.parse(oldCode)); // parse the code (WITH annotations)
```

Comment:
This is where the old `annotate` argument came in handy -- we just parsed without building any annotations, to make sure it was valid code. IIRC, the annotation process was pretty expensive. I saw noticeable jank when dealing with programs >400LOC.